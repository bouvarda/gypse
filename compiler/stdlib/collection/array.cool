
class Array(var length: Int) {
    var buf: Any = alloc(length);
    private def alloc(size: Int): Any = syscall;

    def get(i: Int): Any = syscall;
    def put(i: Int, value: Any): Unit = syscall;
}
