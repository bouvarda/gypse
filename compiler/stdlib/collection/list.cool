
class ListItem(var value: Any, var next: Any) {
    def set_next(item: ListItem): Unit = {
        next = item;
    };
}

class ListIterator(var element: ListItem) {
    def has_next(): Bool = {
        element != null
    };
    def next(): Any = {
        var value: Any = element.value;
        element = element.next;
        value
    };
}

class List() {
    var head: ListItem = null;
    var tail: ListItem = null;
    var length: Int = 0;

    def len(): Int = this.length;

    def push(value: Any): Unit = {
        var item: ListItem = new ListItem(value, null);
        if (head == null) {
           head = item;
           tail = item;
        } else {
            tail.set_next(item);
            tail = item;
        };
        length = length + 1;
    };

    def get(index: Int): Any = {
        if (index >= length) {
            abort("Index out of bounds");
        };
        var i: Int = 0;
        var cur: ListItem = head;
        while (i < index) {
            cur = cur.next;
            i = i + 1;
        };
        cur.value
    };

    def iterator(): ListIterator = {
        new ListIterator(head)
    };
}
