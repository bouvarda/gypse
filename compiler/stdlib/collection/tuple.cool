
class Tuple(var x: Any, var y: Any) {
}