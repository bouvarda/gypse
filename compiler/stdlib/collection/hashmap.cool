import std.collection.List;
import std.collection.Tuple;

class HashMap() {
    var tuples: List = null;
    {
        tuples = new List();
    };

    def put(key: String, value: Any): Unit = {
        tuples.push(new Tuple(key, value));
    };

    def get(key: String): Any = {
        var it: ListIterator = tuples.iterator();
        var result: Any = null;
        while ((result == null) && it.has_next()) {
            var next: Tuple = it.next();
            if (key.equals(next.x)) {
                result = next.y;
            };
        };
        result
    };

    def len(): Int = tuples.length;
}