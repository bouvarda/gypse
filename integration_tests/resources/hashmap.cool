import std.collection.HashMap;

class Main() {
    {
        var colors: HashMap = new HashMap();
        colors.put("red", 42);
        colors.put("green", 56);
        colors.put("blue", 12);

        out_int(colors.len());
        out_nl();
        out_int(colors.get("green"));
        out_nl();
        out_int(colors.get("red"));
        out_nl();
        out_int(colors.get("blue"));
        out_nl();
        out_int(colors.get("yellow"));
    };
}