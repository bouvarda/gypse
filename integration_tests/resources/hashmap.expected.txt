3
56
42
12
0